Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb

```diff
@@ -0,0 +1,205 @@
+    # content is a string
+    # are_more_bytes_available is a boolean stating if the container has more bytes to read
+    def read()
+        raise 'not implemented'
+    end
+end
+
+class JsonParser
+  def initialize(logger, linear_reader)
+    @logger = logger
+    @linear_reader = linear_reader
+    @stream_base_offset = 0
+
+    @stream_reader = StreamReader.new(@logger,@linear_reader)
+    @parser_factory = javax::json::Json.createParserFactory(nil)
+    @parser = @parser_factory.createParser(@stream_reader)
+  end
+
+  def parse(on_json_cbk, on_skip_malformed_cbk)
+    completed = false
+    while !completed
+      completed, start_index, end_index = parse_single_object(on_json_cbk)
+      if !completed
+
+        # if current position in the stream is not a well formed JSON then
+        # I can skip all future chars until I find a '{' so I won't have to create the parser for each char
+        json_candidate_start_index = @stream_reader.find('{',   end_index)
+        json_candidate_start_index = @stream_reader.get_cached_stream_length - 1 if json_candidate_start_index.nil?
+        @logger.debug("JsonParser::parse Skipping Malformed JSON (start: #{start_index} end: #{end_index} candidate: #{json_candidate_start_index - 1}).  Resetting the parser") 
+        end_index = json_candidate_start_index - 1  
+
+        on_skip_malformed_cbk.call(@stream_reader.get_stream_buffer(start_index, end_index))
+        @stream_reader.drop_stream(end_index + 1)
+        @stream_reader.reset_cached_stream_index(0)
+
+        @stream_base_offset = 0
+        @parser.close()
+        if @stream_reader.get_cached_stream_length <= 1
+          on_skip_malformed_cbk.call(@stream_reader.get_stream_buffer(0, -1))
+          return
+        end
+        @parser  = @parser_factory.createParser(@stream_reader)
+      end
+    end
+  end
+  
+  private
+  def parse_single_object(on_json_cbk)
+    depth = 0
+    stream_start_offset = 0
+    stream_end_offset = 0
+    while @parser.hasNext
+      event = @parser.next
+  
+      stream_offset = @parser.getLocation().getStreamOffset()
+  
+  
+      if event == javax::json::stream::JsonParser::Event::START_OBJECT
+        depth = depth + 1
+      elsif event == javax::json::stream::JsonParser::Event::END_OBJECT
+        depth = depth - 1 # can't be negative because the parser handles the format correctness
+  
+        if depth == 0
+          stream_end_offset = @parser.getLocation() .getStreamOffset() - 1
+          @logger.debug ("JsonParser::parse_single_object Json  object found stream_start_offset: #{stream_start_offset} stream_end_offset: #{stream_end_offset}")
+  
+          on_json_cbk.call(@stream_reader.get_stream_buffer(stream_start_offset - @stream_base_offset,  stream_end_offset - @stream_base_offset))
+          stream_start_offset = stream_end_offset + 1
+  
+          #Drop parsed bytes
+          @stream_reader.drop_stream(stream_end_offset  - @stream_base_offset)
```

Comment:
I assume this call is pretty fast and we do this to trim down on working memory usage?

Reply:
yes